[Merged by Bors] - feat(linear_algebra/projectivization/independence): defines (in)dependence of points in projective space

MichaelBlyth opened the pull request:
This PR only provides definitions and basic lemmas. In an upcoming pull request we use this to prove the axioms for an abstract projective space.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [x] depends on: #14714 
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Michael14071407 committed:
Added files

Michael14071407 committed:
added documentation

adamtopaz committed:
Update independence.lean

adamtopaz commented on src/linear_algebra/projective_space/basic.lean:
```diff
@@ -92,6 +92,17 @@ lemma mk_eq_mk_iff (v w : V) (hv : v ≠ 0) (hw : w ≠ 0) :
 quotient.eq'
 
+/-Two nonzero vectors go to the same point in projective space iff one is in
+the span of the other. -/
```
```suggestion
/- Two nonzero vectors go to the same point in projective space if and only if one is a 
scalar multiple of the other. -/
```

Michael14071407 committed:
Fixed abbreviations and wording.

Michael14071407 committed:
Fixed spacing.

adamtopaz commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,109 @@
+
+namespace projectivization
+
+/- A linearly independent familty of nonzero vectors gives an independent family of points
```
```suggestion
/-- A linearly independent family of nonzero vectors gives an independent family of points
```
Docstrings for definitions need to use the syntax `/-- ... -/`. The syntax `/- ...-/` is just a multiline comment.
There is a similar linting error for `dependent` below.

ocfnash replied:
Thanks for working on this @Michael14071407 but note the typo `familty` Adam highlighted above remains.

Also note that the syntax for docstrings does not terminate as `--/` but rather as `-/`.

Michael14071407 committed:
Added appropriate doc strings.

Michael14071407 committed:
Updated the styling of comments on lemmas.

ocfnash committed:
feat(linear_algebra/*): add lemma `linear_independent.finite_of_is_noetherian`

This replaces `fintype_of_is_noetherian_linear_independent` which gave the same
conclusion except demanded `strong_rank_condition R` instead of just `nontrivial R`.

Also some other minor gaps filled.

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,109 @@
+
+/-- A linearly independent familty of nonzero vectors gives an independent family of points
+in the projective space. --/
+inductive independent {ι : Type*} : (ι → ℙ K V) → Prop
```
It's quite rare that we need to use `inductive` and at first glance it doesn't look necessary here. Is there a reason I'm missing?

If not, it looks to me like a better definition would be either:
```lean
def independent {ι : Type*} (f : ι → ℙ K V) := linear_independent K $ projectivization.rep ∘ f
```
(which of course you show is equivalent) or maybe even better:
```lean
def independent {ι : Type*} (f : ι → ℙ K V) := complete_lattice.independent $ λ i, (f i).submodule
```

ocfnash replied:
NB: it's possible #14714 might be relevant

adamtopaz replied:
I think the main benefit of an inductive approach is that by using the `rcases` or `induction` tactic you automatically get representatives for your inductive family.
This also has the benefit that it's easy to construct independence proofs if you already have an independent family of vectors.

At the end of the day, this is equivalent to showing that `projectivization.rep ∘ f` is independent.

If we had an analogue of `quotient.lift_on` for families of elements of quotients as opposed to just 1, 2 or 3 elements, then I think that would be the ideal way to define independence in this case (projectivization is defined as a quotient).

Related zulip discussion:
https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/quotient.2Elift_on_family/near/286096782

All that aside, I do think we should try to include in this PR the equivalence with the lattice-theoretic definition of independence in your second code block!

ocfnash replied:
Thanks @adamtopaz , let's keep the use of `inductive` then.

Regarding the equivalence with the lattice-theoretic definition, this will follow trivially from `independent_iff_linear_independent_of_ne_zero` included in #14714

ocfnash committed:
Fix name collision.

These fixes are necessary because we have introduced the import
`data.finite.basic` in `order.compactly_generated` which is then
transitively exported.

ocfnash committed:
fix

ocfnash committed:
fix

ocfnash committed:
fix

ocfnash committed:
merge master

Michael14071407 committed:
Merge remote-tracking branch 'origin/ocfnash/linear_indep_finite' into projective_dependence

Michael14071407 committed:
Added lattice theoretic definition of independence

Michael14071407 committed:
Simplified proof of equivalence of definitions.

adamtopaz committed:
Merge branch 'master' into projective_dependence

adamtopaz committed:
fix

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,127 @@
+/-- A definition of the independence of a family of points in projective space wherein the family
+is independent if and only if the family of submodules the points determine is independent. -/
+def independent' {ι : Type*} (f : ι → ℙ K V) := complete_lattice.independent $ λ i, (f i).submodule
+
```
I don't think we need this as a definition.
```suggestion
```

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,127 @@
+    refine complete_lattice.independent.linear_independent (projectivization.submodule ∘ f) h _ _,
+    { intro i, rw [function.comp_app projectivization.submodule f i, submodule_eq],
+      exact submodule.mem_span_singleton_self ((f i).rep) },
+    { intro i, rw function.comp_app, exact rep_nonzero (f i) } },
```
This can be golfed a little as follows:
```suggestion
lemma independent_iff_complete_lattice_independent {ι : Type*} (f : ι → ℙ K V) :
  independent f ↔ (complete_lattice.independent $ λ i, (f i).submodule) :=
begin
  refine ⟨_, λ h, _⟩,
  { rintros ⟨f, hf, hi⟩,
    simpa [submodule_mk, complete_lattice.independent_iff_linear_independent_of_ne_zero hf] },
  { rw independent_iff,
    refine h.linear_independent (projectivization.submodule ∘ f) (λ i, _) (λ i, _),
    { simpa only [function.comp_app, submodule_eq] using submodule.mem_span_singleton_self _, },
    { exact rep_nonzero (f i) } },
```

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,127 @@
+/-- A family of points in a projective space are independent if and only if their
+representatives are linearly independent. -/
+lemma independent_iff (ι : Type*) (f : ι → (ℙ K V)) : independent f ↔
+  linear_independent K (projectivization.rep ∘ f) :=
```
```suggestion
lemma independent_iff {ι : Type*} {f : ι → ℙ K V} :
  independent f ↔ linear_independent K (projectivization.rep ∘ f) :=
```

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,127 @@
+-/
+
+
+variables {K V : Type*} [field K] [add_comm_group V] [module K V]
```
If we add `ι` and `f` as a `variables` then we can drop them from all the statements below.
```suggestion
variables {ι K V : Type*} [field K] [add_comm_group V] [module K V] {f : ι → ℙ K V}
```
Note also that since all lemmas about `f` are iff lemmas, it's actually more desirable to have `f` consistently as an implicit variable.

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,127 @@
+lemma independent_iff (ι : Type*) (f : ι → (ℙ K V)) : independent f ↔
+  linear_independent K (projectivization.rep ∘ f) :=
+begin
+  split,
```
If you change this to:
```suggestion
  refine ⟨λ h, _, λ h, _⟩,
```
then you can drop the two `intro h` statements below

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,127 @@
+/-- Dependence is the negation of independence. -/
+lemma dependent_iff_not_independent {ι : Type*} (f : ι → ℙ K V) :
+  dependent f ↔ ¬ independent f :=
+by { rw [dependent_iff, independent_iff] }
```
```suggestion
by rw [dependent_iff, independent_iff]
```

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,127 @@
+/-- Independence is the negation of dependence. -/
+lemma independent_iff_not_dependent {ι : Type*} (f : ι → ℙ K V) :
+  independent f ↔ ¬ dependent f :=
+by { rw [dependent_iff, independent_iff, not_not] }
```
```suggestion
by rw [dependent_iff, independent_iff, not_not]
```

ocfnash reviewed (commented):
Thanks, this is getting really close now.

Michael14071407 committed:
Added the changes suggested by Oliver Nash.

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,117 @@
+lemma independent_iff : independent f ↔ linear_independent K (projectivization.rep ∘ f) :=
+begin
+  refine ⟨λ h, _, λ h, _⟩,
+  { induction h with ff hff hh,
```
```suggestion
  refine ⟨_, λ h, _⟩,
  { rintros ⟨ff, hff, hh⟩,
```

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,117 @@
+  { intro h,
+    convert dependent.mk _ _ h,
+    { ext, simp only [mk_rep] },
+    { intro i, apply rep_nonzero } }
```
```suggestion
  refine ⟨_, λ h, _⟩,
  { rintros ⟨ff, hff, hh1⟩,
    contrapose! hh1,
    choose a ha using λ (i : ι), exists_smul_eq_mk_rep K (ff i) (hff i),
    convert hh1.units_smul a⁻¹,
    ext i,
    simp only [← ha, inv_smul_smul, pi.smul_apply', pi.inv_apply, function.comp_app] },
  { convert dependent.mk _ _ h,
    { ext i, simp only [mk_rep] },
    { exact λ i, rep_nonzero (f i) } }
```

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,117 @@
+by rw [dependent_iff, independent_iff, not_not]
+
+/-- Two points in a projective space are dependent if and only if they are equal. -/
+@[simp] lemma pair_dependent_iff_eq (u v : ℙ K V) : dependent ![u, v] ↔ u = v :=
```
```suggestion
@[simp] lemma dependent_pair_iff_eq (u v : ℙ K V) : dependent ![u, v] ↔ u = v :=
```

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,117 @@
+
+/-- Two points in a projective space are independent if and only if the points are not equal. -/
+@[simp] lemma pair_independent_iff_neq (u v : ℙ K V) : (independent ![u, v]) ↔ u ≠ v :=
+by rw [independent_iff_not_dependent, pair_dependent_iff_eq u v]
```
```suggestion
@[simp] lemma independent_pair_iff_neq (u v : ℙ K V) : independent ![u, v] ↔ u ≠ v :=
by rw [independent_iff_not_dependent, dependent_pair_iff_eq u v]
```

ocfnash commented on src/linear_algebra/projective_space/independence.lean:
```diff
@@ -0,0 +1,117 @@
+
+/-- Independence is the negation of dependence. -/
+lemma independent_iff_not_dependent : independent f ↔ ¬ dependent f :=
+by rw [dependent_iff, independent_iff, not_not]
```
```suggestion
by rw [dependent_iff_not_independent, not_not]
```

ocfnash reviewed (approved):
Please apply the suggested changes and then feel free to merge!

bors d+

Michael14071407 committed:
Added more changes suggested by Oliver.
